nautobot_plugin_nornir/plugins/credentials/nautobot_secrets.py

```
@@ -0,0 +1,46 @@
+"""Credentials class designed to work with Nautobot Secrets Functionality."""
+
+from nautobot.extras.choices import SecretsGroupAccessTypeChoices, SecretsGroupSecretTypeChoices
+from nautobot.extras.models.secrets import SecretsGroupAssociation
+
+from .nautobot_orm import NautobotORMCredentials
+
+
+class CredentialsNautobotSecrets(NautobotORMCredentials):
+    """Credentials Class designed to work with Nautobot Secrets Functionality."""
+
+    def get_device_creds(self, device):
+        """Return the credentials for a given device.
+
+        Args:
+            device (dcim.models.Device): Nautobot device object
+
+        Return:
+            username (string):
+            password (string):
+            secret (string):
+        """
+        if device.secrets_group:
+            self.username = device.secrets_group.get_secret_value(
```

Is there any risk of this returning a `None`? If so, what would happen?

---

It could potentially return an empty string, and could potentially raise one of several exceptions (typically subclasses of `SecretError`). This comes around again to the general question of what this plugin is "expected" to do as far as error handling? From a quick glance it looks like both the SettingsVars and EnvVars classes can potentially return None for any of the values - should this plugin use a try-catch block and set the values to None if an exception occurs?

---

I suppose the true error handling should be done in the plugin that is using this plugin since this is more of a shim layer.  @itdependsnetworks i'm thinking returning None might be a better option?  What are your thoughts.

---

yes, that makes sense, it is valid to use the inventory but not ever need a secret. Which begs the question, should we raise an error if one is not set? I genuinely do not have a strong opinion, but want to be deliberate and document it.

---

After reading @glennmatthews suggestion above that seems to make more sense to me.  For GC for instance we already have handling if the user/pass/secret are not set, so returning None from this provider would follow the current methodology of the other providers.  I'll get it updated.

---

@itdependsnetworks updated to return none(s) if the secret doesn't exist.  Mind a quick re-review?